subprojects/files/src/main/java/org/gradle/internal/file/DefaultFileHierarchySet.java

```diff
@@ -218,14 +254,10 @@ boolean contains(String filePath, int offset) {
             return false;
         }
 
-        public void collect(int depth, List<String> prefixes) {
-            if (depth == 0) {
-                prefixes.add(prefix);
-            } else {
-                prefixes.add(depth + ":" + prefix.replace(File.separatorChar, '/'));
-            }
+        public void collect(int depth, HierarchyConsumer consumer) {
+            consumer.consumePrefix(depth, prefix);
             for (Node child : children) {
-                child.collect(depth + 1, prefixes);
+                child.collect(depth + 1, consumer);
             }
         }
```

Comment:
Both `collect` and `consume` sounds a bit weird to me. This sounds like a visitor pattern to me. Shall we call it that?

Reply:
The parameter name should be `visitor`.

Reply:
Done.